### ISSUE TITLE
Member Search: Add None to Group filter

### ISSUE THREAD

**Issue by bernd-wechner** [tendenci/tendenci#944]
There is a problem with member imports in that they do not support any form of group binding (group_id is not supported, but requested in https://github.com/tendenci/tendenci/issues/934).

This means after import it's a manual step to associate the newly imported members with groups. Alas, there is no easy way to find those members (it's looking like a SQL query my end for now). But a minimal need to facilitate it is to find them.

And one way to do that is if on the People/member search page under the Group filter we added support for "None" namely list all members that have no group. 

I may PR such a fix soon (or not). But I certainly need some way to easily move these new imports into groups.

**Comment by eschipul** [comment]
@bernd-wechner Great suggestion. And I see your point. 

Short term, you can do this immediately now in a two step process using /explorer/  and a subsequent import to a catch all group (see below). The help file is here:
[https://www.tendenci.com/help-files/how-use-django-sql-explorer-tendenci/](https://www.tendenci.com/help-files/how-use-django-sql-explorer-tendenci/)

Ideally groupid 1 is associated with entity id 1 and would be immutable and couldn't be deleted but you can't really enforce that at the application level.

As shown in the help file this sql would give you all interactive users. Hence your suggestion.
```
SELECT  u.first_name, u.last_name, u.email, u.username,
          u.is_staff,  u.is_superuser, p.salutation, p.company, 
          p.position_title, p.phone,  p.address, p.address2, 
          p.member_number, p.city, p.state, p.zipcode,  p.country, 
          p.url, p.sex, p.address_type, p.phone2, p.fax,
          p.work_phone,  p.home_phone, p.mobile_phone, p.notes, 
          p.admin_notes 
      FROM auth_user u  
      INNER JOIN profiles_profile p 
         ON u.id=p.user_id 
      WHERE u.is_active=True  
         AND p.status=True 
         AND p.status_detail='active'
```

The data structure I chose in 1999 when I started it is a true EAV model... and while not all of the software adheres to it this is more of a balance of working with the community and not everyone is as adamant about data models as me. However, I would encourage you to also take a look at the intent of entities (money/reporting/scaling) and groups (organization/security/filtering). Typically groups are a subset of a true entities taxonomy.

That help file (get a cup of coffee for this one as it is quite detailed) is located at:

https://www.tendenci.com/help-files/entities/

And as always, pull requests are greatly appreciated. Especially smaller more incremental ones so it's not a review of "all or nothing". We greatly appreciate you being a part of the community! And I'd be happy to add you to the credits page if you are OK with that?

Ed

**Comment by lashirley21** [comment]
Yes , here https://tendenci.readthedocs.io/en/latest/credits/credits.html :)

**Comment by bernd-wechner** [comment]
Thanks Ed, been busy fixing it and there are a good few PRs for (Jenny I presume) to review, merge in or provide feedback on. All pumped out while working on member imports for this new site. 

Funny you should mention EAV, because the profiles model is anything but, it is such a sparse model with so many fields unused I am sure by many sites implementing Tendenci the table is so wide it's hard to do anything with manually in a table editor and might have been a candidate fro EAV, though to be honest, canonical Django probably wouldn't implement a profiles table that way. 

Re: credits, thank you, much appreciated. Very happy to be listed. I am likely to fall a tad quit after a busy period with member imports, but maybe will be pottering around in the coming year with enhancements for the site I'm implementing. Something I predict because Tendenci IMHO is a diamond in the rough, and has all the hallmarks of a passion product with low funding and organizational backing over its history.

I came to it looking for a way to divest this clubs of its use of spreadsheets for membership management. Primarily because privacy legislation makes storing member data on cloud drive in aspreadhseet (where it's synced to any number of past committee members local drives) a bit fo a shambles. Of course a database is just one part of a privacy solution but an important one.

So I first conducted a review of options and when I spotted tendenci as a Django based FOSS platform with large user base and mature and long in the tooth, it was for me (as a Django programmer of sorts) a no brainer to trial. The trial passed muster and now we're implementing.

Our demo site for now sits here: http://new.hogs.org.au/ but we have to migrate before new year.

**Comment by eschipul** [comment]
@bernd-wechner I definitely enjoy table top games. Chess would be my favorite, Dominoes (the "real" game which is actually called "Dominoes" - to the point that people now sometimes call it "straight dominoes" because of the variations. Simple rules, but a fast moving game that is far more complex than people realize.) Not sure if darts and billiards/pool fall into that category. But I digress.

I just added you to credits. Follow up comment on technical stuff.

**Comment by eschipul** [comment]
@bernd-wechner  Tendenci was originally called "codebase" and I started in 1999. Written in ASP using a mySQL database. Written from the framework on up (remember COM objects). But you can't really be open source with Microsoft so we moved to the Django framework which has evolved and there remain philosophical differences. 

Given Django is after all, their framework, we have to bend to doing things the Django way. "functions" in "views.py", "decorators", "template tags". The auth_user object having four fields and leaving extensibility to the developer leads to inconsistencies across projects, etc...  - I love the framework but _it is different_.

I noticed a lot of your pull requests are related to "groups" and "perms" which is a particularly challenging area. My intent with groups was slightly different from Django's, so there are a lot of variations in how these function. Be careful in this area in particular given the thousands of people counting on security. 


Tendenci is EAV in Entities in particular. But given the constraint on compute resources when we started much of it was modeled on mainframe flat files to make SQL easier. I don't like security divided into tSQL (in Microsoft lingo) for example. I prefer security all at the application layer for clarity. And with reinforcement at OS and the cloud layer. 

So yes, there are quite a few divergences. And while EAV is great for programmers, it's not easy for regular people who just want to run a dashboard. It's an intricate balance. (And when we did all of this, there was no JSON or noSQL or BigTable or Data Lakes. And XML/XSLT was just tooooooo slow. We worked with what we had.)

Back to data structure:

I can't emphasize enough, and I repeat a lot, that
Entities = money (invoices can only have one (1) entityid) - this is for committees to budget etc. And entities are the heirarchy.)
Groups - are for security and organizational purposes. Groups are bound to entities. Not all entities have a group however. 

In production everything on our cloud is in docker containers. Working on the change over to serverless. I'm testing that with terraform on gcp initially. 

And yes, @jennyq is amazing and is our technical lead for sure! She's brilliant.

**Comment by bernd-wechner** [comment]
Thanks for the history! 

Didn't touch perms intentionally, that would have been part of the riskiest and biggest PR I offered. Interestingly I documented the utility of "groups" yesterday as:

- A link to list people in that group
- An email address to reach members of that group
- segregation of permissions (access to data)

As I'm importing members piecewise (started with 1, then 2 then 5 then 10, and am at 20(as my learning and confidence on impacts improves I can do larger lots and soon the whole lot ;-) I grouped members for the above reasons based on role (primarily committee/member) and memberships status (lot of overlap there with membership_type and expire_dt).

In doing that I tweaked the quick member slit to provide a useful overview of that progress (it was a fairly lean list with just name and company so added a few useful columns and sorting). In doing that I found the group_set attribute auth.user. Went down a rabbit hole of reading and testing on that. A (provisional) learning on Django 2.2 at least (I use 3.0 elsewhere) is that:

- group_set worked as documented and intended as an attribute of a auth.user object. Tick. 
- It would not work (caused Django to bail) when used as a field specifier (specifically on members, "user__group_set" or "user__group_set__name")

On inspection auto.user had three group_set candidates: creator, owner and members. The former two had explicit related_names the last one not.

Django parses ORM attributes and field references differently (different code paths) 

Field references (strings) support __ as a relation follower, while attributes don't (Python syntax handles the relationship tree)

So the provisional learning is that Field reference syntax (against all the documentation) seems not to support the implicit related names Django creates (group_set) in this case, but does support explicit related_names.

Given IMHO it's best practice in any case to use explicit related_names I did gamble the refactor of group_set to user_groups. I think I got it (I did a search, seek, destroy on group_set in the whole code base anyhow and confirmed visually that it was a auth.user. group_set reference and patched it.

It's a gambit because I don't have or see a solid regression testing option. I can see what models were affected in the commit:

- base/template_tags/base_filters
- committees/views
- events/managers
- memberships/forms
- memberships/managers
- perms/managers
- perms/utils
- studygroups/views

So not a small set, but not huge, and yes touching perms as you feared (in four spots).

And sop yes, int he commit message I wrote to self (and not sure what travelled on the PR) that this warrants a regression test on all those apps anyhow (at least, but ideally a comprehensive one) before it is ever released. I have only tested the areas I'm working with (far from comprehensive).
